File: src/demo_python_topic/demo_python_topic/turtle_control.py
import rclpy
from rclpy.node import Node#导入rclpy库
from geometry_msgs.msg import Twist#导入geometry_msgs消息类型
from turtlesim.msg import Pose#导入turtlesim消息类型
from math import atan2, fabs, sqrt, pow#导入数学库
#Pose 告诉我们"海龟在哪里"
#Twist 告诉海龟"要怎么走"

class TurtleControlNode(Node):
    def __init__(self,node_name):
        super().__init__(node_name)
        self.publisher_ = self.create_publisher(Twist,"turtle1/cmd_vel",10)#创建一个发布者
        self.subscriber_ = self.create_subscription(Pose,"turtle1/pose",self.pose_callback,10)#创建一个订阅者
        self.target_x_ = 1.0#目标位置x
        self.target_y_ = 1.0#目标位置y
        self.k_ = 1.0#比例系数
        self.v_max_ = 3.0#最大速度

    def pose_callback(self,pose):
        print("当前位置:x=%f,y=%f",pose.x,pose.y)
        #1.计算当前位置与目标位置的距离差和角度差
        distance = sqrt(pow(self.target_x_-pose.x,2)+pow(self.target_y_-pose.y,2))#计算距离差
        angle = atan2((self.target_y_-pose.y),(self.target_x_-pose.x))-pose.theta#计算角度差
        #2.控制策略：计算速度和角速度
        msg = Twist()# Twist用于发送运动控制命令
        if distance > 0.1:#如果距离差大于0.1，则进行控制
            if fabs(angle) > 0.2:#如果角度差大于0.2，则进行转向
                msg.angular.z = fabs(angle)  # 控制转向
            else:
                msg.linear.x = self.k_*distance  # 控制前进
        #3.限制线速度最大值
        if msg.linear.x > self.v_max_:
            msg.linear.x = self.v_max_
        #4.发布控制命令
        self.publisher_.publish(msg)

def main(args=None):
    rclpy.init(args=args)#初始化ros2
    node = TurtleControlNode('turtle_control')#创建节点
    rclpy.spin(node)#运行节点
    rclpy.shutdown()#关闭ros2

File: src/demo_cpp_topic/src/turtle_control.cpp
#include "rclcpp/rclcpp.hpp"//导入rclcpp库
#include "geometry_msgs/msg/twist.hpp"//导入geometry_msgs消息类型
#include <chrono>//导入时间库
#include "turtlesim/msg/pose.hpp"//导入turtlesim消息类型
using namespace std;
using namespace std::chrono;

class TurtleControlNode : public rclcpp::Node  // 继承rclcpp::Node
{
private:  
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_; // 创建一个发布者
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscriber_; // 创建一个订阅者
    double target_x_ = 1.0;//目标位置x
    double target_y_ = 1.0;//目标位置y
    double k_=1.0;//比例系数
    double v_max_ = 3.0;//最大速度
    
public:
    explicit TurtleControlNode(const string& node_name):Node(node_name)// 构造函数
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);// 创建一个发布者
        subscriber_ = this->create_subscription<turtlesim::msg::Pose>("turtle1/pose",10,
        bind(&TurtleControlNode::pose_callback,this,placeholders::_1));
    }

    void pose_callback(const turtlesim::msg::Pose::SharedPtr pose)//回调函数：计算速度和角速度
    {
        //1.获取当前位置
        auto current_x = pose->x;
        auto current_y = pose->y;
        
        RCLCPP_INFO(get_logger(),"当前位置:x=%f,y=%f",current_x,current_y);
        //2.计算当前位置与目标位置的距离差和角度差
        auto distance = sqrt(
            pow(target_x_-current_x,2)+
            pow(target_y_-current_y,2)
        );
        // atan2(y,x)计算的是目标点相对于当前点的方位角(弧度)
        // (target_y_-current_y)和(target_x_-current_x)分别是目标点和当前点的y、x方向差值
        auto angle = atan2((target_y_-current_y),(target_x_-current_x))-pose->theta;
        
        //3.控制策略：计算速度和角速度
        auto msg = geometry_msgs::msg::Twist();
        if(distance>0.1)
        {
            if(fabs(angle)>0.2)
            {
                msg.angular.z = fabs(angle);
            }
            else{
                msg.linear.x =k_*distance;
            }
        }

        //4.限制线速度最大值
        if(msg.linear.x>v_max_)
        {
            msg.linear.x = v_max_;
        }
        //5.发布速度和角速度
        RCLCPP_INFO(get_logger(),"线速度:x=%f,角速度:z=%f",msg.linear.x,msg.angular.z);
        publisher_->publish(msg);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc,argv);
    auto node = make_shared<TurtleControlNode>("turtle_control_node");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}